### PR TITLE
Let solver integrate flows

### PR DESCRIPTION
**Pull request by SouthEndMusic**
Fixes https://github.com/Deltares/Ribasim/issues/1431.

To do:
- [x] Let the solver integrate edge flows
- [x] Let the solver integrate basin forcings
- [x] Let the solver integrate flows for allocation input
- [x] Let the solver integrate realized user demand flows
- [x] Update Jacobian sparsity and related test
- [x] Docstrings etc. (maybe some parts of the documentation that now only mention the storages and the PID integral term as states)

### CODE REVIEW

**Comment by SouthEndMusic** [comment]
There seems to be something broken in adaptive timestepping. It looks like timesteps are allowed to be arbitrarily small (no matter the solver settings), which would explain why running the `backwater` model takes a ridiculous amount of time and this doesn't throw an error:

https://github.com/Deltares/Ribasim/blob/79aa2dfdc3e586c66e38317358dcea02bbe9a73b/core/test/bmi_test.jl#L14-L15

It also looks like that comment is outdated, because you don't give a negative `dt` but I suspect a too small one

**Comment by visr** [comment]
> It also looks like that comment is outdated

It doesn't look outdated to me. First you update to t=dt0, then to t=dt2/0, which is before the current t. The SciML methods expect a dt argument, whereas BMI expects a t argument.

File: core/src/model.jl [hunk 1]
```diff
@@ -129,17 +133,26 @@ function Model(config::Config)::Model
...
     @debug "Created callbacks."
 
+    # Only have finite tolerance on storage states
+    abstol = copy(u0)
+    reltol = copy(u0)
+    abstol .= Inf
+    reltol .= Inf
+    abstol.storage .= config.solver.abstol
+    reltol.storage .= config.solver.reltol
+
```
Comment (SouthEndMusic) [inline core/src/model.jl:155]: If desired, we could apply different tolerances to different components of the states this way

**Review by visr — COMMENTED** [review]
Looks good, though the "mean_flow" test perhaps needs some double checking to see if it is ok. A flow of 0.84 where it should be 1 seems quite a large difference, though if it is only for tiny dt and evens out properly over time it should be fine.

File: core/src/solve.jl [hunk 2]
```diff
@@ -597,7 +597,41 @@ function formulate_flow!(
...
+)::Nothing
+    (; basin, allocation, user_demand) = p
+
+    # Flows over edges
+    flow = get_tmp(graph[].flow, storage)
+    du.flow_integrated .= flow
+
+    # Basin forcings
+    vertical_flux = get_tmp(basin.vertical_flux, storage)
+    forcings_integrated(du) .= vertical_flux
```
[the remark this answers is not in the document]
Comment (SouthEndMusic) [reply core/src/solve.jl:614]: I've added a test

File: core/src/solve.jl [hunk 3]
```diff
@@ -597,7 +597,41 @@ function formulate_flow!(
...
+            get_influx(basin, edge[1])
+        else
+            get_flow(graph, edge..., 0)
+        end
+    end
+
+    # Realized user demand
+    for (i, id) in enumerate(user_demand.node_id)
+        du.realized_user_demand_bmi[i] = get_flow(graph, inflow_id(graph, id), id, storage)
+    end
```
[the remark this answers is not in the document]
Comment (SouthEndMusic) [reply core/src/solve.jl:629]: Should we build in an error within `inflow_id` etc. to check whether it is requested for a node type for which the neighbors are precalculated?

**Review by evetion — CHANGES_REQUESTED** [review]
Can't comment on the solver changes/setup, but overall looks ok to me, some small comments.

FWIW I would be useful for others to have a nice PR description (which turns into a commit message after (squash) merging) at least detailing what changed in the internal API for other developers who didn't see these changes.

**Comment by evetion** [comment]
Also, this seems a net change in terms of lines added/removed, contrary to the original idea in #1431. Do we think this lowered our maintenance? How is the performance affected?

**Comment by SouthEndMusic** [comment]
So it turns out that this is pretty bad for performance, it basically negates all performance gain from https://github.com/Deltares/Ribasim/pull/1457 as measured iwith the HWS model. So here's a wild idea: What if we set up a parallel ODE problem for the flow integration with an explicit solver?

**Comment by visr** [comment]
Too bad this didn't work out performance wise, good effort though.

**Pull request closed without merging**